src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs

```diff
@@ -178,22 +195,32 @@ private async Task<bool> ProcessDwToHwTransaction(Guid operationId, StellarBase.
                 CreatedAt = DateTime.UtcNow
             };
 
+            // save without state to prevent changing of tx hash
+            await _broadcastRepository.InsertOrReplaceAsync(broadcast);
+
+            _chaos.Meow(nameof(ProcessDwToHwTransaction));
+
             var assetId = _blockchainAssetsService.GetNativeAsset().Id;
-            if (await _balanceRepository.DecreaseBalanceAsync(assetId, fromAddress, hash, amount))
-            {
-                broadcast.State = TxBroadcastState.Completed;
-            }
-            else
+            var balance = await _balanceRepository.GetAsync(assetId, fromAddress);
+
+            if (balance.Balance < amount)
             {
                 broadcast.State = TxBroadcastState.Failed;
                 broadcast.Error = "Not enough balance!";
                 broadcast.ErrorCode = TxExecutionError.NotEnoughBalance;
             }
+            else
+            {
+                await _balanceRepository.RecordOperationAsync(assetId, fromAddress, updateLedger, 0, hash, (-1) * amount);
```

Comment:
I suggest to use operationId.GetHash() instead of 0 for operationId parameter. And looks like better to use calculate some kind of real hash for DW-HW transaction according to its parameters. Cause operationId and hash should be idempotent for the same DW-HW to make RecordOperationAsync work properly.

Reply:
We need unique transaction identifier and unique identifier of action within transaction (called operationId in Stellar), so
- changed hash to our guid-ed OperationId, as it uniquely identifies the transaction, just without dashes to look more hash-y;
- while we have only single action within DW->HW transaction, we can use any value as action identifier, changed to numeric hash of our guid-ed OperationId to add more diversity;
- removed unnecessary ChaosKitty calls;
- removed unnecessary broadcast data checks and arguments;
- added comments for future;